### DISCUSSION TITLE
v0.0.14

### DISCUSSION THREAD

**Discussion by erogol** [coqui-ai/TTS#492] [Versions & Releases]
# :frog: v0.0.14
## 🐞Bug Fixes
- [x] Remove breaking line from Tacotron models. (👑 @a-froghyra)

## 💾 Code updates
- [x] **BREAKING:** Coqpit integration for config management and the first 🐸TTS recipe, for LJSpeech Check #476.

Every model now tied to a Python class that defines the configuration scheme. It provides a better interface and lets the user know better what are the default values, expected value types, and mandatory fields.

Specific model configs are defined under `TTS/tts/configs` and `TTS/vocoder/configs`. `TTS/config/shared_configs.py` hosts configs that are shared by all the :frog: TTS models.  Configs shared by `tts` models are hosted under `TTS/tts/configs/shared_configs.py` and shared by `vocoder` models are under `TTS/vocoder/configs/shared_config.py`.

For example `TacotronConfig` follows `BaseTrainingConfig -> BaseTTSConfig -> TacotronConfig`.
- [x] **BREAKING:** Remove `phonemizer` support due to License conflict. 

This essentially deprecates the support for all the models using phonemes as input. Feel free to suggest in-place options if you are affected by this change. 

- [x] Start hosting :woman_cook: **recipes** under :frog: TTS. The first recipe is for Tacotron2-DDC with LJspeech dataset under `TTS/recipes/`. 

Please check [here](https://github.com/coqui-ai/TTS/tree/main/recipes) for more details.
- [x] Add `extract_tts_spectrograms.py` that supports GlowTTS and Tacotron1-2. (👑 @Edresson)
- [x] Add `version.py` (👑 @chmodsss)

<hr /><em>This discussion was created from the release <a href='https://github.com/coqui-ai/TTS/releases/tag/v0.0.14'>v0.0.14</a>.</em>

**Comment by george-roussos** [comment]
Regarding the removal of `phonemizer` and replacement suggestions, I can think of an option, which is using the [G2P function](https://montreal-forced-aligner.readthedocs.io/en/latest/g2p.html) of [Montreal Forced Aligner](https://montreal-forced-aligner.readthedocs.io/en/latest/).

The basic requirement is a pronunciation lexicon that will be used for the task (languages such as English and German have very big ones that also happen to be open source). It is an option that is quite stable in my personal opinion (and observation), straightforward to train and much quicker to generate pronunciations for unknown words. You can also choose to train models on different n-gram orders and call them based on word lengths.

The pipeline would be as follows:

1. Perform a lexicon lookup to acquire pronunciation for each word
2. If the word is there, grab the pronunciation and synthesize
3. If not, use the g2p model to generate a probable pronunciation

The results to get are much better than `phonemizer`, because this way we get to implement stress markers (which `phonemizer` does not support), have more fine-grained control over what model will be used (thus improve accuracy) and also have better results (since `phonemizer` performs rule-based guessing and in many languages these rules change depending on phenomena such as compounding). In terms of speed, while yes, guessing pronunciations for unknown words does not happen instantly, using a transducer to perform the task is faster than instantiating `phonemizer` every single time and in the case where the text to synthesize has all the words in the pronunciation lexicon, the look-up is very fast.

**Reply by erogol** [reply]
I checked the MFA before but to me the limitations are;

- No number or acronym normalization.
- No contextual G2P. Meaning, it does not handle words pronounces differently depending on the context. 

Do you think MFA can provide these? 

I guess, using raw phonemes without the normalization only solves the words that are hard to pronounce like `echo`. Would you agree? 

Can you also expand on how MFA helps us to have better control over the speech? By the lexicon lookup?

**Reply by george-roussos** [reply]
I agree that `MFA` does not provide contextual `G2P`, but `phonemizer` doesn't either, or does it? Do you mean contextual as in picking the correct pronunciation for homographs? In any case, if contextual is what we are looking for (which I think is a different topic, although very much necessary), we should be looking into tagging the lexicon and by extension the text (so then we have the information for cross-referencing available) using different means.

WRT the acronym normalization, `MFA` does not do that either, but these can be handled during the text normalisation stage (and then the normalised text may be used for `G2P`). So I guess what I am trying to say is that we can leverage MFA for what it offers and extend the normalization functionality of Coqui.

For the last point: I was on about how we can implement things like stress markers and so further reduce the amount of errors, especially in languages with primary and secondary stresses. Also, since we are able to use multiple MFA models (depending on the length of the word), it is much more likely that we will get the correct phonemic sequence, while the rule-based guessing of `phonemizer` is much more fixed.

**Reply by erogol** [reply]
Maybe MFA is not the optimal solution but can be a good start :) 

Do you think all that can be taught to a NN? Maybe we can train a Seq2Seq model using the language-specific lexicon as input.

MFA uses n-gram models for G2P for unknown words, right? 

In the doc they also say their method does not work well for non-phonemic languages like English if I remember correctly. 

Some other libraries I checked so far for any interest 

(I am expanding the list as we find more alternatives)

https://github.com/dmort27/epitran
https://github.com/CUNY-CL/wikipron
https://github.com/rhasspy/gruut

**Reply by george-roussos** [reply]
Awesome!

On the `seq2seq` objective, I was actually going to mention it but did not -- I tried it a while ago, but, personally, the results I got were quite sub-optimal. There were mistakes in the transcription and I found the model quite prone to overfitting, probably because I did not do enough hyperparameter tuning. This may be a problem, since it may imply that for every language they have to be different. Also, the transcription time on CPU was slow. In the end, I abandoned it, because I did not notice any improvements. I will try to dig up the name of the library I used (it was a while ago and I am forgetting).

`MFA` use n-gram models for G2P, yeah. I am also seeing that their WER for some languages (including English) are higher; it makes sense yeah, because of the fact that English is less transparent orthography-wise. But I think it is definitely worth a try, seeing as there is a big pronunciation lexicon for English available (so many words already present) and there is a pretrained model for English, too.

**Comment by mbarnig** [comment]
I have some experience with TTS from the past when I explored espeak, MaryTTS and Festival to build a luxembourgish TTS voice. Last year I published a book in french about the history of speech synthesis. I am relatively new in the domain of machine-learning. I revisited my old ideas concerning the creation of an lb-TTS-voice a few weeks ago when Coqui-AI was launched. At the same time I discovered the [rhasspy/larynx](https://github.com/rhasspy/larynx) project developed by Michael Hansen (alias synesthesiam). Larynx uses [gruut]( https://github.com/rhasspy/gruut) to transform text into phonemes.

I wonder why there is no close collaboration between coqui-tts and larynx/gruut. I think both Coqui-AI and Rhasspy are outstanding and I would like to congratulate and thank the authors of both projects.

Greetings from Luxembourg,

Marco Barnig

**Reply by erogol** [reply]
We used to collaborate with @rhasspy and actually he (if I remember correctly) created our docker image but I don't know why he gave up on that. Maybe he can jump into the discussion. I forgot about gruut and it looks like a good candidate.

**Reply by mbarnig** [reply]
Thank you for your answer.

You are right, synesthesiam created a [coqui-docker file](https://github.com/synesthesiam/coqui-docker) 7 days ago. I use it for my training trials with coqui tacotron2-dcc.

Concerning gruut I think it's worth to explore it as a replacement-candidate for espeak & phonemizer.

Rhasspy/Gruut is a tokenizer, text cleaner, and IPA phonemizer for several human languages. Pretrained models are available for different languages. Gruut allows to transform raw text into phonetic pronunciations, similar to phonemizer. Unlike phonemizer, Gruut looks up words in a pre-built lexicon (pronunciation dictionary) or guesses word pronunciations with a pre-trained grapheme-to-phoneme model.

Gruut uses four files in a folder for each language:

* phonemes.txt : list of used phonemes
* lexicon.db : SQlite pronunciation dictionary
* g2p.fst : grapheme to phoneme model
* language.yml : configuration 

Gruut uses the following tools to setup the four language files:

* [Babel](https://pypi.org/project/Babel/)
* [Num2Words](https://pypi.org/project/num2words/)
* [gruut-IPA](https://github.com/rhasspy/gruut-ipa)
* [phonetisaurus-pypi](https://github.com/rhasspy/phonetisaurus-pypi)

Creating the gruut language files for luxembourgish was very straightforward and easy and I recommend the [rhasspy/gruut](https://github.com/rhasspy/gruut) tool.

An interesting additional feature of gruut is to convert the IPA-phonemes in indexes to simplify the voice models for training and inference.

**Reply by erogol** [reply]
I just saw that Num2Words is LGPL so I'm not sure if it is compatible with MIT which is the license of gruut. I guess this is the most important blocker for now.

**Reply by mbarnig** [reply]
I don't have any practical experience with software licenses. To refresh my knowledge I read the contribution [Over 10% of Python Packages on PyPI are Distributed Without Any License](https://snyk.io/blog/over-10-of-python-packages-on-pypi-are-distributed-without-any-license/) on the snykblog.

I will contact synesthesiam to ask his position about the license mismatch between gruut and num2words.

**Reply by mbarnig** [reply]
I just got aware that the license problem was already raised in April 2020 in the [Mozilla discourse](https://discourse.mozilla.org/t/gpl-in-the-way-license-issues-regarding-current-source-code/58820).

**Reply by domcross** [reply]
gruut has some dependencies that prevents pip-install on a number of platforms among them Raspbian, MacOS, Tegra (Nvidia Jetson/Xavier), ...

**Reply by erogol** [reply]
@mbarnig  yes unidecode needs to go too. Thanks for pointing the thread

**Reply by erogol** [reply]
@domcross good to know that. It is definitely a limiting factor.

**Reply by synesthesiam** [reply]
@erogol if I were to write a wrapper around `espeak-ng` with a similar API to phonemizer and an MIT license, would that work for you license-wise?

I'd be happy to get gruut into a state too where it could help out here. It has limited language support, but I think more people could contribute with the MIT license and fairly simple pipeline (look up word/phoneme pairs in SQLite database, optionally with part-of-speech tags). There are also some [interesting discussions](https://github.com/rhasspy/larynx/issues) for German and French speakers :slightly_smiling_face:  

Regarding num2words, it is under the LGPL which (according to Wikipedia):

> The license allows developers and companies to use and integrate a software component released under the LGPL into their own (even proprietary) software without being required by the terms of a strong copyleft license to release the source code of their own components. However, any developer who modifies an LGPL-covered component is required to make their modified version available under the same LGPL license. 

I have made sure to release [our modifications](https://github.com/rhasspy/num2words) under the same license (support for Swedish and Farsi).

**Reply by erogol** [reply]
@synesthesiam  unfortunately wouldn't work since GPL makes all the code written around it GPL. We need a no espeak solution.

LGPL seems fine. Sorry about the comment earlier. 

I think gruut still the best possible solution. 

@synesthesiam do you think there is anything we can do about @domcross's comment?

**Reply by mbarnig** [reply]
@erogol. I suggest another tool as a candidate to eventually replace phonemizer : [DeepPhonemizer](https://github.com/as-ideas/DeepPhonemizer), launched a month ago by Christian Schäfer  @cschaefer26. It's MIT licensed. I experienced some troubles with the code, but I think it's an interesting concept because a lot of languages, scraped from [Wikipron](https://github.com/CUNY-CL/wikipron), can be supported.

Personally I am still in favor of gruut because the concept is very elegant and didactical.

**Reply by kdavis-coqui** [reply]
@mbarnig DeepPhonemizer is MIT and it claims, as far as I can tell, to have models that are also MIT.

However, the models are trained off of data (e.g. cmudict-ipa, cmudict, and wikipron) that are not MIT. So the MIT license for the models may not be correct.

In particular:
- **cmudict-ipa** doesn't have a license which means it models trained from it can not be licensed as MIT
- **cmudict** seems to have it's own [special license](https://github.com/cmusphinx/cmudict/blob/master/LICENSE) which looks like a slightly modified BSD two clause. So in all likelihood models trained from it can be licensed as MIT.
- **wikipron** [wikipron data](https://github.com/CUNY-CL/wikipron/tree/master/data) from which the models are made seem to be under a [Wiktionary license](https://en.wiktionary.org/wiki/Wiktionary:Copyrights) which seems to imply that the MIT license on the models isn't correct as the Wiktionary license requires if you use the Wiktionary materials, then "your materials have to be licensed under the same, similar, or compatible license (to the Wiktionary license)", which doesn't appear to be true for the MIT license.

**Reply by synesthesiam** [reply]
@domcross Can you be more specific? [Phonetisaurus](https://github.com/AdolfVonKleist/Phonetisaurus) is what I assume you're referring to, which relies on pre-compiled binaries. I have few possibilities:

1. Create a Python library to do g2p inference using the pre-trained Phonetisaurus models (started, but not fully working)
2. Switch to a system that uses [python-crfsuite](https://github.com/scrapinghub/python-crfsuite) (working, not as accurate as Phonetisaurus)
3. Switch to a numpy-only system that's based on [this Apache 2.0 project](https://github.com/Kyubyong/g2p) (working, not as accurate as option 2)

gruut also uses Babel to do locale-specific number/currency parsing, but this could be made optional if it's a blocker.

@erogol I'll start cleaning up gruut's API and test it out with a fork of :frog: TTS

**Reply by domcross** [reply]
> Can you be more specific? [Phonetisaurus](https://github.com/AdolfVonKleist/Phonetisaurus) is what I assume you're referring to, which relies on pre-compiled binaries.

Exactly, I am refering to the Phonetisaurus dependency. Option 1) sounds most promising to me ;-)

**Reply by mbarnig** [reply]
@kdavis-coqui. This license issue is complicated and tricky. My understanding now is that in all cases we need pronunciation dictionaries with the correct license, whether to use them in gruut or in DeepPhonemizer, to build a language model with a suitable license for Coqui-TTS. I googled for free pronunciation dictionaries and found only sources with a license mismatch. It seems to me that most people don't worry about the license restrictions.

I used the luxembourgish pronunciation dictionaries [lb-lu.dic](https://github.com/spellchecker-lu/dictionary-lb-lu) and [phonetic transcription spellchecker-lu.txt]( https://github.com/PeterGilles/Luxembourgish-language-resources) to build the grapheme to phoneme model g2p.fst with gruut. These dictionaries are licensed under the European Union Public Licence ([EUPL version 1.1](https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12)). In the appendix it's stated that compatible licences according to article 5 EUPL are:
```   
- GNU General Public License (GNU GPL) v. 2
- Open Software License (OSL) v. 2.1, v. 3.0
- Common Public License v. 1.0
- Eclipse Public License v. 1.0
- Cecill v. 2.0
```   
It's not yet clear to me if this EUPL licence allow me to use or to share the luxembourgish gruut phoneme model g2p.fst.

**Reply by synesthesiam** [reply]
Most of my IPA lexicons have come from here: https://github.com/open-dict-data/ipa-dict which claims an MIT license. The Credits section mentions [this cmudict-ipa](https://github.com/lingz/cmudict-ipa) as a source for English, which **does** show an MIT license.

**Reply by kdavis-coqui** [reply]
@synesthesiam Generally in machine learning one has to consider at least 3 licenses

- License of the data 
- License of the model trained on the data
- License of the code that uses the model

The license of the data dictates the possible model licenses and the model license dictates the possible code licenses.

For your case, cmudict-ipa looks good, with MIT, but the addition of stress markers using syllabify is problematic as the syllabify repo does not specify any license, despite the link from ipa-dict saying syllabify is MIT.

If a repo doesn't specify a license, "no one may reproduce, distribute, or create derivative works from" the repo[[1](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/licensing-a-repository#choosing-the-right-license)]. So the modified version of cmudict-ipa is not really MIT. But it appears as if the non-modified version of cmudict-ipa is MIT.

As you can see, this can get painful.

**Reply by kdavis-coqui** [reply]
@mbarnig When you say
>
> It seems to me that most people don't worry about the license restrictions.

It's definitely true and makes this all the more difficult, and it's already hard even if people didn't make this (understandable) mistake.

As to the EUPL, I don't have any experience with that license. So I can't say anything too helpful.

However, the EUPL does state that

> If the Licensee Distributes or Communicates Derivative Works or copies thereof based upon both
> the Work and another work licensed under a Compatible Licence, this Distribution or Communication can be done
> under the terms of this Compatible Licence

So it looks like when combining something under EUPL with one of the Compatible Licences you have a clear path forward as to how to license the combined work.

**Comment by thorstenMueller** [comment]
Probably a stupid idea, but should we ask the maintainer of espeak whether he would be willing to adjust the license for espeak?

**Reply by kdavis-coqui** [reply]
@thorstenMueller It's not a stupid idea.

However, the maintainer can't simply re-license a repo. The maintainer would have to get permission from every contributor to the repo to re-license the repo. This usually proves impractical for repos with many contributors.

**Reply by thorstenMueller** [reply]
Okay, thanks for the explanation @kdavis-coqui. So that will probably not happen.

**Comment by cschaefer26** [comment]
@kdavis-coqui Thx for pointing out the license issues, I probably will have to add a disclaimer to the phonemizer repo. Hovewer, if you are interested I could train a model on an open source dataset and make it available so you can test it out.

**Comment by mbarnig** [comment]
@kdavis-coqui and other contributors. Thank you very much for your detailed explanations about the licenses. 
The topic becomes progressively clear to me. I have learned so much the last days with your great communities,
not only about licenses, but about machine-learning in general, about coding and about sharing ideas.

**Comment by synesthesiam** [comment]
@kdavis-coqui Do you think lexicons derived from Wiktionary would have any licensing issues?

**Reply by kdavis-coqui** [reply]
@synesthesiam In the licensing terms[[1](https://foundation.wikimedia.org/wiki/Terms_of_Use/en#7._Licensing_of_Content)] it states that

> **Modifications or additions to material that you re-use:** When modifying or making additions to text that you have obtained from a Project website, you agree to license the modified or added content under CC BY-SA 3.0 or later (or, as explained above, another license when exceptionally required by the specific Project edition or feature).

However, there are exceptions under "fair use", see section _g_ of the section linked above. I'd guess a lawyer would be best equipped to determine these exceptions.

**Comment by synesthesiam** [comment]
I've created [a pull request](https://github.com/coqui-ai/TTS/pull/523) to re-enable phoneme-based TTS models using [gruut](https://github.com/rhasspy/gruut) :slightly_smiling_face:
